clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -2517,6 +2580,40 @@ class SyclKernelBodyCreator : public SyclKernelFieldHandler {
     return true;
   }
 
+  // Default inits the type, then calls the init-method in the body
+  void handleSyclKernelHandlerType(ParmVarDecl *KernelHandlerArg) {
+
+    // Create local clone of kernel handler
+    KernelHandlerClone = createKernelHandlerClone(
+        SemaRef.getASTContext(), DeclCreator.getKernelDecl(), KernelHandlerArg);
+
+    // Default initialize clone
+    InitializedEntity VarEntity =
+        InitializedEntity::InitializeVariable(KernelHandlerClone);
+    InitializationKind InitKind =
+        InitializationKind::CreateDefault(KernelCallerSrcLoc);
+    InitializationSequence InitSeq(SemaRef, VarEntity, InitKind, None);
+    ExprResult Init = InitSeq.Perform(SemaRef, VarEntity, InitKind, None);
+    KernelHandlerClone->setInit(
+        SemaRef.MaybeCreateExprWithCleanups(Init.get()));
+    KernelHandlerClone->setInitStyle(VarDecl::CallInit);
+
+    // Add declaration statement to openCL kernel body
+    Stmt *DS =
+        new (SemaRef.Context) DeclStmt(DeclGroupRef(KernelHandlerClone),
+                                       KernelCallerSrcLoc, KernelCallerSrcLoc);
+    BodyStmts.push_back(DS);
+
+    // Generate init call
+    // FIXME: Should this be restricted to targets which do not have native
+    // support for specialization constants?
+    const auto *RecordDecl =
+        KernelHandlerClone->getType()->getAsCXXRecordDecl();
+    // FIXME: This call generates __init function bound to kernel object clone.
+    // Fix this.
+    createSpecialMethodCall(RecordDecl, InitSpecConstantsBuffer, BodyStmts);
```

Comment:
I realized this generated incorrect call while writing the test. I will correct this in next patch